hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestAbfsPaginatedDelete.java

```diff
@@ -0,0 +1,279 @@
+  }
+
+  @Override
+  public void setup() throws Exception {
+    isHnsEnabled = this.getConfiguration().getBoolean(FS_AZURE_TEST_NAMESPACE_ENABLED_ACCOUNT, false);
+    loadConfiguredFileSystem();
+    super.setup();
+    this.superUserFs = getFileSystem();
+    this.firstTestUserGuid = getConfiguration()
+        .get(FS_AZURE_BLOB_FS_CHECKACCESS_TEST_USER_GUID);
+
+    if(isHnsEnabled) {
+      // setting up ACL permissions for test user
+      setFirstTestUserFsAuth();
+      setDefaultAclOnRoot(this.firstTestUserGuid);
+    }
+  }
+
+  /**
+   * Test to check that recursive deletePath works with paginated enabled and
+   * disabled for both empty and non-empty directory.
+   * When enabled appropriate xMsVersion should be used.
+   * @throws Exception
+   */
+  @Test
+  public void testRecursiveDeleteWithPagination() throws Exception {
+    testRecursiveDeleteWithPaginationInternal(false, true, DECEMBER_2019_API_VERSION);
+    testRecursiveDeleteWithPaginationInternal(false, true, AUGUST_2023_API_VERSION);
+    testRecursiveDeleteWithPaginationInternal(false, false, DECEMBER_2019_API_VERSION);
+    testRecursiveDeleteWithPaginationInternal(false, false, AUGUST_2023_API_VERSION);
+    testRecursiveDeleteWithPaginationInternal(true, true, DECEMBER_2019_API_VERSION);
+    testRecursiveDeleteWithPaginationInternal(true, false, AUGUST_2023_API_VERSION);
+  }
+
+  /**
+   * Test to check that non-recursive delete works with both paginated enabled
+   * and disabled only for empty directories.
+   * Pagination should not be set when recursive is false.
+   * @throws Exception
+   */
+  @Test
+  public void testNonRecursiveDeleteWithPagination() throws Exception {
+    testNonRecursiveDeleteWithPaginationInternal(true);
+    testNonRecursiveDeleteWithPaginationInternal(false);
+  }
+
+  /**
+   * Test to check that with pagination enabled, invalid CT will fail
+   * @throws Exception
+   */
+  @Test
+  public void testRecursiveDeleteWithInvalidCT() throws Exception {
+    testRecursiveDeleteWithInvalidCTInternal(true);
+    testRecursiveDeleteWithInvalidCTInternal(false);
+  }
+
+  public void testRecursiveDeleteWithPaginationInternal(boolean isEmptyDir, boolean isPaginatedDeleteEnabled,
+      String xMsVersion) throws Exception {
+    final AzureBlobFileSystem fs = isHnsEnabled ? this.firstTestUserFs : getFileSystem();
+    TracingContext testTracingContext = getTestTracingContext(this.firstTestUserFs, true);
+    Path testPath;
+    if (isEmptyDir) {
+      testPath = new Path("/emptyPath" + StringUtils.right(
+          UUID.randomUUID().toString(), 10));
+      fs.mkdirs(testPath);
+    } else {
+      testPath = createSmallDir();
+    }
+
+    // Set the paginated enabled value and xMsVersion at client level.
+    AbfsClient client = ITestAbfsClient.setAbfsClientField(
+        fs.getAbfsStore().getClient(), "xMsVersion", xMsVersion);
+    client.getAbfsConfiguration().setIsPaginatedDeleteEnabled(isPaginatedDeleteEnabled);
```

Comment:
Lets isolate this configSet to a given thread. Other thread in parallel should not be affected.